// @coinbase/agentkit/dist/action-providers/x402/schemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FetchPaymentInfoSchema = exports.PaidRequestSchema = void 0;
const zod_1 = require("zod");
exports.PaidRequestSchema = zod_1.z
    .object({
    url: zod_1.z.string().url().describe("The URL of the x402-protected API endpoint"),
    method: zod_1.z
        .enum(["GET", "POST", "PUT", "DELETE", "PATCH"])
        .default("GET")
        .describe("The HTTP method to use for the request"),
    headers: zod_1.z.record(zod_1.z.string()).optional().describe("Optional headers to include in the request"),
    body: zod_1.z.any().optional().describe("Optional request body for POST/PUT/PATCH requests"),
})
    .strip()
    .describe("Instructions for making a paid request to an x402-protected API");
exports.FetchPaymentInfoSchema = zod_1.z
    .object({
    url: zod_1.z.string().url().describe("The URL of the x402-protected API endpoint"),
    method: zod_1.z
        .enum(["GET", "POST", "PUT", "DELETE", "PATCH"])
        .default("GET")
        .describe("The HTTP method to use for the request"),
    headers: zod_1.z.record(zod_1.z.string()).optional().describe("Optional headers to include in the request"),
})
    .strip()
    .describe("Instructions for fetching payment information from an x402-protected API endpoint");
